x/programs/rust/sdk-macros/src/lib.rs

```diff
@@ -343,6 +346,86 @@ pub fn state_keys(_attr: TokenStream, item: TokenStream) -> TokenStream {
     .into()
 }
 
+#[derive(Debug)]
+struct KeyPair {
+    key_comments: Vec<Attribute>,
+    key_vis: Visibility,
+    key_type_name: Ident,
+    key_fields: Fields,
+    value_type: Ident,
+}
+
+impl Parse for KeyPair {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let key_comments = input.call(Attribute::parse_outer)?;
+        let key_vis = input.parse::<Visibility>()?;
+        let key_type_name: Ident = input.parse()?;
+        let lookahead = input.lookahead1();
+
+        // TODO: fail on named fields
+        let key_fields = if lookahead.peek(token::Paren) {
+            let fields = input.parse()?;
+            Fields::Unnamed(fields)
+        } else {
+            Fields::Unit
+        };
+
+        input.parse::<Token![=>]>()?;
+        let value_type: Ident = input.parse()?;
+
+        Ok(Self {
+            key_comments,
+            key_vis,
+            key_type_name,
+            key_fields,
+            value_type,
+        })
+    }
+}
+
+#[proc_macro]
+pub fn state_schema(input: TokenStream) -> TokenStream {
+    // parse out the key-pairs
+    let key_pairs =
+        parse_macro_input!(input with Punctuated::<KeyPair, Token![,]>::parse_terminated);
+    key_pairs
+        .into_iter()
+        .enumerate()
+        .map(|(i, val)| (i as u8, val))
+        .fold(
+            quote! {},
+            |mut token_stream,
+             (
+                i,
+                KeyPair {
+                    key_comments,
+                    key_vis,
+                    key_type_name,
+                    key_fields,
+                    value_type,
+                },
+            )| {
+                token_stream.extend(Some(quote! {
+                    #(#key_comments)*
+                    #[derive(Copy, Clone, bytemuck::NoUninit)]
```

Comment:
What does the error message look like if `bytemuck` isn't imported? What if it *is* imported, but is missing the `derive` feature?